u5-rest-apis/features/covid-controller.js

```diff
@@ -0,0 +1,126 @@
+const { db } = require('../_services/firebase-admin-initialized')
+const { firestore } = require('firebase-admin')
+
+const readCovids = async (req, res) => {
+  try {
+    // 1. Inputs
+    // none
+
+    // 2. Query
+    const query = db.collection('covid-latest')
+      .get()
+
+    // 3. Response
+    const payload = (await query)
+      .docs
+      .map(doc => doc.data())
+      .map(({ date, stateId, stateName, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }) =>
+        ({ date: date.toMillis(), stateId, stateName, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }))
+
+    res.json({
+      result: 'ok',
+      payload: payload,
+      count: payload.length
+    })
+  } catch (err) {
+    console.error(err)
+    res.status(500).json({
+      result: 'error',
+      payload: [],
+      count: 0
+    })
+  }
+}
+
+const readCovid = async (req, res) => {
+  try {
+    // 1. Inputs
+    const stateId = req.params.stateId.toUpperCase()
+
+    // 2. Query
+    const query = db.collection('covid-history')
+      .doc(stateId)
+      .get()
+
+    // 3. Response
+    const snapshot = await query
+    if (!snapshot.exists) { return res.status(404).json({ result: 'not found' }) }
+
+    const { stateName, history } = snapshot.data()
+    const payload = {
+      stateId,
+      stateName,
+      history: history.map(item => ({
+        date: item.date.toMillis(),
+        cases: item.cases,
+        casesNew: item.casesNew,
+        vaccineOne: item.vaccineOne,
+        vaccineOnePercent: item.vaccineOnePercent,
+        vaccineComplete: item.vaccineComplete,
+        vaccineCompletePercent: item.vaccineCompletePercent
+      }))
+    }
+    res.json({ result: 'ok', payload })
+  } catch (err) {
+    console.error(err)
+    res.status(500).json({ result: 'error' })
+  }
+}
+
+const createCovid = async (req, res) => {
+  try {
+    // 1. Inputs
+    const { stateId, date, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent } = req.body
+    const record = { stateId, date: firestore.Timestamp.fromMillis(date), cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }
+
+    // 2. Query
+    const query = db.collection('covid-latest')
+      .doc(stateId)
+      .set(record, { merge: true })
+
+    // 3. Response
+    await query
+    res.sendStatus(201)
+  } catch (err) {
+    console.error(err)
+    res.sendStatus(500)
+  }
+}
+
+const updateCovid = async (req, res) => {
+  try {
+    // 1. Inputs
+    const stateId = req.params.stateId.toUpperCase()
+    const { date, cases, casesNew } = req.body
+    const record = { date: firestore.Timestamp.fromMillis(date), cases, casesNew }
+
+    // 2. Query
+    const query = db.collection('covid-latest').doc(stateId).set(record, { merge: true })
```

Comment:
หมายเหตุ: สำหรับห้องปฏิบัติการ คุณควรถือว่า `date` ในคำขอเท่ากับ `date` ใน `covid-latest` (นี่คือการอัปเดตข้อมูล "ล่าสุด") โซลูชันที่สมจริงยิ่งขึ้นจะตรวจสอบว่าสิ่งนี้เป็นจริงโดยใช้ธุรกรรม -- ตัวอย่างเช่น การอัปเดตข้อมูลสำหรับวันที่เก่าไม่ควรเปลี่ยน `covid-latest` มันจะอัพเดทแค่ `covid-history`

ในส้อมของฉัน ฉันได้แบ่งปันทั้งวิธีแก้ปัญหาง่ายๆ (สำหรับแล็บ) และวิธีแก้ปัญหาที่สมจริงยิ่งขึ้น (ซึ่งซับซ้อนกว่ามาก)

-----

Note: For the lab you should assume that `date` is equal to the `date` already in `covid-latest` (this is updating the "latest" data). A more realistic solution would check that this is true using a transaction -- for example updating data for an old date should not change `covid-latest`; it would only update `covid-history`.

In my fork I have shared both the simple solution for the lab and a more realistic solution (which is considerably more complex).